File: aws/cdk/typescript/lib/stacks/ecs-stack.ts
import { NestedStack, RemovalPolicy } from "aws-cdk-lib";
import {
  Cluster,
  FargateTaskDefinition,
  FargateService,
  ContainerImage,
  Protocol,
  LogDrivers,
  ContainerDependencyCondition,
  Secret as EcsSecret,
} from "aws-cdk-lib/aws-ecs";
import { Vpc, Subnet, SecurityGroup } from "aws-cdk-lib/aws-ec2";
import { Secret } from "aws-cdk-lib/aws-secretsmanager";
import { Role } from "aws-cdk-lib/aws-iam";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { Construct } from "constructs";
import { ECSstackProps } from "../interfaces/ecs-interface";

export class ECSstack extends NestedStack {
  constructor(scope: Construct, id: string, props: ECSstackProps) {
    super(scope, id, props);

    const {
      name,
      description,
      vpcId,
      availabilityZones,
      subnets,
      securityGroups,
      ecsTaskRoleArn,
      task,
      locations,
      privatePackage,
      enterpriseCloud,
    } = props;

    const {
      cpu = 1024,
      memory = 3072,
      image = "gatlingcorp/control-plane:latest",
      command = [],
      secrets = {},
      environment = {},
      cloudwatchLogs = true,
    } = task;

    const volumeNname: string = "control-plane-conf";
    const path: string = "/app/conf";

    const configContent: string = `
      {
        "control-plane": {
          token = \${?CONTROL_PLANE_TOKEN},
          description: ${description},
          enterprise-cloud: ${JSON.stringify(enterpriseCloud)},
          locations: ${JSON.stringify(
            locations.map((location) => ({
              type: "aws",
              ...location,
            }))
          )},
          repository: ${JSON.stringify(privatePackage ? { type: "aws", ...privatePackage } : {})}
        }
      }
      `;

    const vpc = Vpc.fromVpcAttributes(this, "ExistingVpc", {
      vpcId: vpcId,
      availabilityZones: availabilityZones,
      privateSubnetIds: subnets,
    });

    const cluster = new Cluster(this, "GatlingEnterpriseCluster", {
      clusterName: `${name}-cluster`,
      vpc,
    });

    const taskRole = Role.fromRoleArn(this, "EcsTaskRole", ecsTaskRoleArn);

    const logGroup = cloudwatchLogs
      ? new LogGroup(this, "LogGroup", {
          logGroupName: `/ecs/${name}-service`,
          retention: RetentionDays.ONE_MONTH,
          removalPolicy: RemovalPolicy.DESTROY,
        })
      : undefined;

    const taskDefinition = new FargateTaskDefinition(
      this,
      "GatlingTaskDefinition",
      {
        family: `${name}-task`,
        taskRole,
        executionRole: taskRole,
        cpu: cpu,
        memoryLimitMiB: memory,
        volumes: [
          {
            name: volumeNname,
          },
        ],
      }
    );

    const initContainer = taskDefinition.addContainer(
      "ConfLoaderInitContainer",
      {
        image: ContainerImage.fromRegistry("busybox"),
        essential: false,
        environment: {
          CONFIG_CONTENT: configContent,
        },
        command: [
          "/bin/sh",
          "-c",
          `echo "$CONFIG_CONTENT" > ${path}/control-plane.conf`,
        ],
        readonlyRootFilesystem: false,
        logging: cloudwatchLogs
          ? LogDrivers.awsLogs({
              logGroup,
              streamPrefix: "init",
            })
          : undefined,
      }
    );

    initContainer.addMountPoints({
      sourceVolume: volumeNname,
      containerPath: path,
      readOnly: false,
    });

    const ecsSecrets: { [key: string]: EcsSecret } = {};
    if (secrets) {
      for (const [key, secretArn] of Object.entries(secrets)) {
        if (secretArn) {
          const secret = Secret.fromSecretCompleteArn(
            this,
            "ImportedSecret",
            secretArn
          );
          ecsSecrets[key] = EcsSecret.fromSecretsManager(secret);
        }
      }
    }

    const controlPlaneContainer = taskDefinition.addContainer(
      "ControlPlaneContainer",
      {
        image: ContainerImage.fromRegistry(image),
        essential: true,
        secrets: ecsSecrets,
        environment: environment,
        command: command,
        workingDirectory: path,
        portMappings: privatePackage
          ? [
              {
                containerPort: privatePackage?.server?.port
                  ? privatePackage?.server?.port
                  : 8080,
                hostPort: privatePackage?.server?.port
                  ? privatePackage?.server?.port
                  : 8080,
                protocol: Protocol.TCP,
              },
            ]
          : [],
        logging: cloudwatchLogs
          ? LogDrivers.awsLogs({
              logGroup,
              streamPrefix: "main",
            })
          : undefined,
      }
    );

    controlPlaneContainer.addMountPoints({
      sourceVolume: volumeNname,
      containerPath: path,
      readOnly: true,
    });

    controlPlaneContainer.addContainerDependencies({
      container: initContainer,
      condition: ContainerDependencyCondition.SUCCESS,
    });

    new FargateService(this, "GatlingEnterpriseService", {
      serviceName: `${name}-service`,
      cluster,
      taskDefinition,
      desiredCount: 1,
      assignPublicIp: true,
      vpcSubnets: {
        subnets: subnets.map((subnetId) =>
          Subnet.fromSubnetId(this, subnetId, subnetId)
        ),
      },
      securityGroups: securityGroups.map((sgId) =>
        SecurityGroup.fromSecurityGroupId(this, `SecurityGroup-${sgId}`, sgId)
      ),
    });
  }
}

File: aws/cdk/java/src/main/java/com/gatlingenterprise/stacks/ECSstack.java
package com.gatlingenterprise.stacks;

import com.gatlingenterprise.records.controlPlaneProps.ECSstackProps;
import com.gatlingenterprise.records.locationProps.EnterpriseCloud;
import com.gatlingenterprise.utils.ControlPlaneConfigGenerator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import software.amazon.awscdk.NestedStack;
import software.amazon.awscdk.RemovalPolicy;
import software.amazon.awscdk.services.ec2.IVpc;
import software.amazon.awscdk.services.ec2.SecurityGroup;
import software.amazon.awscdk.services.ec2.Subnet;
import software.amazon.awscdk.services.ec2.SubnetSelection;
import software.amazon.awscdk.services.ec2.Vpc;
import software.amazon.awscdk.services.ec2.VpcAttributes;
import software.amazon.awscdk.services.ecs.*;
import software.amazon.awscdk.services.iam.IRole;
import software.amazon.awscdk.services.iam.Role;
import software.amazon.awscdk.services.logs.LogGroup;
import software.amazon.awscdk.services.logs.RetentionDays;
import software.amazon.awscdk.services.secretsmanager.ISecret;
import software.constructs.Construct;

public class ECSstack extends NestedStack {
  public ECSstack(Construct scope, String id, ECSstackProps props) {
    super(scope, id, props);

    String vpcId = props.vpcId();
    List<String> availabilityZones = props.availabilityZones();
    List<String> subnetIds = props.subnetIds();
    List<String> securityGroupIds = props.securityGroupIds();
    String name = props.name();
    Map<String, ISecret> secrets = props.secrets();
    String desc = props.desc();
    String ecsTaskRoleArn = props.ecsTaskRoleArn();
    String image = props.image();
    Map<String, String> environment = props.environment();
    List<String> command = props.command();
    var locations = props.locations();
    var privatePackage = props.privatePackage();
    boolean cloudWatchLogs = props.cloudWatchLogs();
    EnterpriseCloud enterpriseCloud = props.enterpriseCloud();

    String volumeName = "control-plane-conf";
    String path = "/app/conf";

    ControlPlaneConfigGenerator configGenerator =
        new ControlPlaneConfigGenerator(desc, locations, privatePackage, enterpriseCloud);
    String configContent = configGenerator.generateConfig();

    IVpc vpc =
        Vpc.fromVpcAttributes(
            this,
            "ExistingVpc",
            VpcAttributes.builder()
                .vpcId(vpcId)
                .availabilityZones(availabilityZones)
                .privateSubnetIds(subnetIds)
                .build());

    Cluster cluster =
        new Cluster(
            this,
            "GatlingEnterpriseCluster",
            ClusterProps.builder().clusterName(name + "-cluster").vpc(vpc).build());

    IRole taskRole = Role.fromRoleArn(this, "EcsTaskRole", ecsTaskRoleArn);

    LogGroup logGroup =
        cloudWatchLogs
            ? LogGroup.Builder.create(this, "LogGroup")
                .logGroupName("/ecs/" + name + "-service")
                .retention(RetentionDays.ONE_MONTH)
                .removalPolicy(RemovalPolicy.DESTROY)
                .build()
            : null;

    FargateTaskDefinition taskDefinition =
        FargateTaskDefinition.Builder.create(this, "GatlingTaskDefinition")
            .family(name + "-task")
            .taskRole(taskRole)
            .executionRole(taskRole)
            .cpu(1024)
            .memoryLimitMiB(3072)
            .volumes(List.of(Volume.builder().name(volumeName).build()))
            .build();

    ContainerDefinition initContainer =
        taskDefinition.addContainer(
            "ConfLoaderInitContainer",
            ContainerDefinitionOptions.builder()
                .image(ContainerImage.fromRegistry("busybox"))
                .essential(false)
                .environment(Map.of("CONFIG_CONTENT", configContent))
                .command(
                    List.of(
                        "/bin/sh",
                        "-c",
                        "echo \"$CONFIG_CONTENT\" > "
                            + path
                            + "/control-plane.conf; echo \"$CONFIG_CONTENT\""))
                .readonlyRootFilesystem(false)
                .logging(
                    cloudWatchLogs
                        ? LogDrivers.awsLogs(
                            AwsLogDriverProps.builder()
                                .logGroup(logGroup)
                                .streamPrefix("init")
                                .build())
                        : null)
                .build());

    initContainer.addMountPoints(
        MountPoint.builder().sourceVolume(volumeName).containerPath(path).readOnly(false).build());

    Map<String, Secret> ecsSecrets =
        secrets != null
            ? secrets.entrySet().stream()
                .filter(entry -> entry.getValue() != null)
                .collect(
                    Collectors.toMap(
                        Map.Entry::getKey, entry -> Secret.fromSecretsManager(entry.getValue())))
            : Map.of();

    ContainerDefinition controlPlaneContainer =
        taskDefinition.addContainer(
            "ControlPlaneContainer",
            ContainerDefinitionOptions.builder()
                .image(ContainerImage.fromRegistry(image))
                .essential(true)
                .secrets(ecsSecrets)
                .environment(environment)
                .command(command)
                .workingDirectory(path)
                .portMappings(
                    privatePackage != null
                        ? List.of(
                            PortMapping.builder()
                                .containerPort(
                                    privatePackage.server() != null
                                            && privatePackage.server().port() != null
                                        ? privatePackage.server().port()
                                        : 8080)
                                .hostPort(
                                    privatePackage.server() != null
                                            && privatePackage.server().port() != null
                                        ? privatePackage.server().port()
                                        : 8080)
                                .protocol(Protocol.TCP)
                                .build())
                        : List.of())
                .logging(
                    cloudWatchLogs
                        ? LogDrivers.awsLogs(
                            AwsLogDriverProps.builder()
                                .logGroup(logGroup)
                                .streamPrefix("main")
                                .build())
                        : null)
                .build());

    controlPlaneContainer.addMountPoints(
        MountPoint.builder().sourceVolume(volumeName).containerPath(path).readOnly(true).build());

    controlPlaneContainer.addContainerDependencies(
        ContainerDependency.builder()
            .container(initContainer)
            .condition(ContainerDependencyCondition.SUCCESS)
            .build());

    FargateService.Builder.create(this, "GatlingEnterpriseService")
        .serviceName(name + "-service")
        .cluster(cluster)
        .taskDefinition(taskDefinition)
        .desiredCount(1)
        .assignPublicIp(true)
        .vpcSubnets(
            SubnetSelection.builder()
                .subnets(
                    subnetIds.stream()
                        .map(subnetId -> Subnet.fromSubnetId(this, "Subnet-" + subnetId, subnetId))
                        .toList())
                .build())
        .securityGroups(
            securityGroupIds.stream()
                .map(sgId -> SecurityGroup.fromSecurityGroupId(this, "SecurityGroup-" + sgId, sgId))
                .toList())
        .build();
  }
}
